crates/bevy_render/src/mesh/allocator.rs

```diff
@@ -0,0 +1,1025 @@
+/// however, a slot may consist of 2 or even 4 elements. This occurs when the
+/// size of an element isn't divisible by [`COPY_BUFFER_ALIGNMENT`] (4). When we
+/// resize buffers, we perform GPU-to-GPU copies to shuffle the existing
+/// elements into their new positions, and such copies must be on
+/// [`COPY_BUFFER_ALIGNMENT`] boundaries. Slots solve this problem by
+/// guaranteeing that the size of an allocation quantum is divisible by both the
+/// size of an element and [`COPY_BUFFER_ALIGNMENT`], so we can relocate it
+/// freely.
+#[derive(Clone, Copy, PartialEq, Eq, Hash)]
+struct ElementLayout {
+    /// Either a vertex or an index.
+    class: ElementClass,
+
+    /// The size in bytes of a single element (vertex or index).
+    size: u64,
+
+    /// The number of elements that make up a single slot.
+    ///
+    /// Usually, this is 1, but it can be different if [`ElementLayout::size`]
+    /// isn't divisible by 4. See the comment in [`ElementLayout`] for more
+    /// details.
+    elements_per_slot: u32,
+}
+
+/// The location of an allocation and the slab it's contained in.
+struct MeshAllocation {
+    /// The ID of the slab.
+    slab_id: SlabId,
+    /// Holds the actual allocation.
+    slab_allocation: SlabAllocation,
+}
+
+/// An allocation within a slab.
+#[derive(Clone)]
+struct SlabAllocation {
+    /// The actual [`Allocator`] handle, needed to free the allocation.
+    allocation: Allocation,
+    /// The number of slots that this allocation takes up.
+    slot_count: u32,
+}
+
+/// Holds information about all slabs scheduled to be allocated or reallocated.
+#[derive(Default, Deref, DerefMut)]
+struct SlabsToReallocate(HashMap<SlabId, SlabToReallocate>);
+
+/// Holds information about a slab that's scheduled to be allocated or
+/// reallocated.
+#[derive(Default)]
+struct SlabToReallocate {
+    /// Maps all allocations that need to be relocated to their positions within
+    /// the *new* slab.
+    allocations_to_copy: HashMap<AssetId<Mesh>, SlabAllocation>,
+}
+
+impl Default for MeshAllocatorSettings {
+    fn default() -> Self {
+        Self {
+            // 1MB
+            min_slab_size: 1024 * 1024,
+            // 512MB
+            max_slab_size: 1024 * 1024 * 512,
+            // 256MB
+            large_threshold: 1024 * 1024 * 256,
+            // 1.5× growth
+            growth_factor: 1.5,
+        }
+    }
+}
+
+impl Display for SlabId {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        self.0.fmt(f)
+    }
+}
+
+impl Plugin for MeshAllocatorPlugin {
+    fn build(&self, app: &mut App) {
+        let Some(render_app) = app.get_sub_app_mut(RenderApp) else {
+            return;
+        };
+
+        render_app
+            .init_resource::<MeshAllocatorSettings>()
+            .add_systems(
+                Render,
+                allocate_and_free_meshes
+                    .in_set(RenderSet::PrepareAssets)
+                    .before(prepare_assets::<GpuMesh>),
```

Comment:
Why before here, and not after? This stands out as very weird to me.

Reply:
It's because `prepare_assets` clears the contents of the `ExtractedAssets<RenderMesh>` resource after processing it. Because `allocate_and_free_meshes` needs to use that data, it has to run first. (We don't want to merge `allocate_and_free_meshes` with `RenderMesh::prepare_asset` because, among other reasons, that would result in a lot of useless copying when a bunch of meshes load all at the same time and the buffer grows multiple times in a single frame, which happens a lot.)

Reply:
Am I correct that this this system runs on a 1-frame delay then?

Reply:
No. The order in a single frame is this:

1. `extract_render_asset` populates `ExtractedAssets<RenderMesh>` during the extract phase.
2. `allocate_and_free_meshes` examines the `ExtractedAssets<RenderMesh>` to perform the allocations during the asset prep phase.
3. `prepare_asset` processes the `ExtractedAssets<RenderMesh>` again, and clears it out.

So `extract_render_asset` produces the data, `allocate_and_free_meshes` inspects the data, and finally `prepare_asset` consumes the data. This pipeline happens entirely during the frame.

Reply:
Ah ok, thanks for the explanation. This makes sense to me, and mimics how I have it setup in my own code.